File: src/aside/java/org/taktik/icure/logic/impl/PrincipalLogicImpl.java
/*
 * Copyright (C) 2018 Taktik SA
 *
 * This file is part of iCureBackend.
 *
 * iCureBackend is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as published by
 * the Free Software Foundation.
 *
 * iCureBackend is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with iCureBackend.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.taktik.icure.logic.impl;

import com.google.common.reflect.TypeToken;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.taktik.icure.constants.Roles;
import org.taktik.icure.dao.RoleDAO;
import org.taktik.icure.entities.Property;
import org.taktik.icure.entities.PropertyType;
import org.taktik.icure.entities.Role;
import org.taktik.icure.entities.base.Principal;
import org.taktik.icure.entities.embed.Permission;
import org.taktik.icure.logic.PrincipalLogic;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

public abstract class PrincipalLogicImpl<P extends Principal> implements PrincipalLogic<P> {
	protected final Logger log = LoggerFactory.getLogger(getClass());

	protected final TypeToken<P> typeToken = new TypeToken<P>(getClass()) {};

	// Bit indexes
	protected static int DEPENDENCY_INCLUDE_DIRECT = 1 << 1;
	protected static int DEPENDENCY_INCLUDE_HERITED = 1 << 2;

	protected RoleDAO roleDAO;

	protected Class<P> getPrincipalClass() {
		return (Class<P>) typeToken.getRawType();
	}

	protected Set<Role> getParents(Principal principal) {
		return roleDAO.getSet(principal.getParents());
	}


	@Override
	public Set<Property> getProperties(String principalId, boolean includeDirect, boolean includeHerited, boolean includeDefault) {
		Principal principal = getPrincipal(principalId);
		if (principal != null) {
			return buildProperties(principal, includeDirect, includeHerited, includeDefault, new HashSet<>());
		}

		return new HashSet<>();
	}

	@Override
	public Set<Permission> getPermissions(String principalId, String virtualHostId, boolean includeDirect, boolean includeHerited, boolean includeDefault) {
		return null;
	}

	protected Set<Property> buildProperties(Principal principal, boolean includeDirect, boolean includeHerited, boolean includeDefault, Set<PropertyType> ignoredPropertyTypes) {
		log.trace("buildProperties() : principal={}({}), includeDirect={}, includeHerited={}, includeDefault={}", principal.getClass().getSimpleName(), principal.getId(),  includeDirect, includeHerited, includeDefault);

		// Prepare set of properties
		Set<Property> properties = new HashSet<>();

		if (includeDirect) {
			// First add properties directly linked to the principal
			Set<Property> principalProperties = principal.getProperties();
			for (Property p : principalProperties) {
				if (!ignoredPropertyTypes.contains(p.getType())) {
					ignoredPropertyTypes.add(p.getType());
					properties.add(p);
				}
			}
		}

		if (includeHerited) {
			// Get the parent roles, sorted by natural order
			List<Role> parentRolesSorted = new ArrayList<>(getParents(principal));
			Collections.sort(parentRolesSorted, (r1, r2) -> r1.getName().compareToIgnoreCase(r2.getName()));

			// Add properties directly linked to the parents
			for (Role parent : parentRolesSorted) {
				Set<Property> parentProperties = buildProperties(parent, true, false, false, ignoredPropertyTypes);
				properties.addAll(parentProperties);
			}

			// Add properties herited from grand parents
			for (Role parent : parentRolesSorted) {
				Set<Property> parentProperties = buildProperties(parent, false, true, false, ignoredPropertyTypes);
				properties.addAll(parentProperties);
			}
		}

		if (includeDefault) {
			// Get the default role and add property if not overridden in child role
			Role defaultRole = roleDAO.getByName(Roles.DEFAULT_ROLE_NAME);
			if (defaultRole!=null) {
				Set<Property> defaultProperties = defaultRole.getProperties();
				for (Property defaultProp : defaultProperties) {
					if (!ignoredPropertyTypes.contains(defaultProp.getType())) {
						properties.add(defaultProp);
					}
				}
			}
		}

		return properties;
	}

	@Override
	public Set<Role> getAscendantRoles(String principalId) {
		Principal principal = getPrincipal(principalId);
		if (principal != null) {
			return buildAscendantRoles(principal, new HashSet<>());
		}

		return null;
	}

	protected Set<Role> buildAscendantRoles(Principal principal, Set<Role> ignoredRoles) {
		Set<Role> roles = new HashSet<>();
		if (principal != null) {
			// Add this role to ignore list
			if (principal instanceof Role) {
				ignoredRoles.add((Role) principal);
			}

			// Process parents
			if (principal.getParents() != null) {
				for (Role parent : getParents(principal)) {
					if (parent != null) {
						if (!ignoredRoles.contains(parent)) {
							roles.add(parent);
							roles.addAll(buildAscendantRoles(parent, ignoredRoles));
						}
					}
				}
			}
		}

		return roles;
	}

	@Autowired
	public void setRoleDAO(RoleDAO roleDAO) {
		this.roleDAO = roleDAO;
	}
}

File: src/main/kotlin/org/taktik/icure/asynclogic/impl/PrincipalLogicImpl.kt
/*
 *  iCure Data Stack. Copyright (c) 2020 Taktik SA
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU Affero General Public License as
 *     published by the Free Software Foundation, either version 3 of the
 *     License, or (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful, but
 *     WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *     Affero General Public License for more details.
 *
 *     You should have received a copy of the GNU Affero General Public
 *     License along with this program.  If not, see
 *     <https://www.gnu.org/licenses/>.
 */
package org.taktik.icure.asynclogic.impl

import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.asFlow
import kotlinx.coroutines.flow.collect
import kotlinx.coroutines.flow.emitAll
import kotlinx.coroutines.flow.emptyFlow
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.flow.onEach
import kotlinx.coroutines.flow.toList
import org.slf4j.LoggerFactory
import org.taktik.icure.asyncdao.GenericDAO
import org.taktik.icure.asyncdao.RoleDAO
import org.taktik.icure.asynclogic.AsyncSessionLogic
import org.taktik.icure.asynclogic.PrincipalLogic
import org.taktik.icure.constants.Roles
import org.taktik.icure.entities.Role
import org.taktik.icure.entities.base.PropertyStub
import org.taktik.icure.entities.base.PropertyTypeStub
import org.taktik.icure.entities.security.Permission
import org.taktik.icure.entities.security.Principal

abstract class PrincipalLogicImpl<P : Principal>(protected val roleDAO: RoleDAO, protected val sessionLogic: AsyncSessionLogic) : GenericLogicImpl<P, GenericDAO<P>>(sessionLogic), PrincipalLogic<P> {

	protected val log = LoggerFactory.getLogger(javaClass)

	protected fun getParents(principal: Principal): Flow<Role> = flow {
		emitAll(roleDAO.getEntities(principal.getParents()))
	}

	override fun getProperties(principalId: String, includeDirect: Boolean, includeHerited: Boolean, includeDefault: Boolean): Flow<PropertyStub> = flow {
		val principal: Principal? = getPrincipal(principalId)
		principal?.let { emitAll(buildProperties(principal, includeDirect, includeHerited, includeDefault, mutableSetOf<PropertyTypeStub>())) }
			?: emitAll(emptyFlow<PropertyStub>())
	}

	override fun getPermissions(principalId: String, virtualHostId: String, includeDirect: Boolean, includeHerited: Boolean, includeDefault: Boolean): Flow<Permission> = flow {
		emitAll(emptyFlow<Permission>())
	}

	protected fun buildProperties(principal: Principal, includeDirect: Boolean, includeHerited: Boolean, includeDefault: Boolean, ignoredPropertyTypes: MutableSet<PropertyTypeStub>): Flow<PropertyStub> = flow {
		log.trace("buildProperties() : principal={}({}), includeDirect={}, includeHerited={}, includeDefault={}", principal.javaClass.simpleName, principal.id, includeDirect, includeHerited, includeDefault)
		// Prepare set of properties
		val properties = mutableSetOf<PropertyStub>()
		if (includeDirect) { // First add properties directly linked to the principal
			for (p in principal.properties) {
				if (p.type != null && !ignoredPropertyTypes.contains(p.type)) {
					ignoredPropertyTypes.add(p.type)
					properties.add(p)
				}
			}
		}
		if (includeHerited) { // Get the parent roles, sorted by natural order
			val parentRolesSorted = getParents(principal).toList().sortedWith(compareBy<Role, String>(String.CASE_INSENSITIVE_ORDER) { r -> r.name ?: "" })
			// Add properties directly linked to the parents
			for (parent in parentRolesSorted) {
				val parentProperties = buildProperties(parent, true, false, false, ignoredPropertyTypes).toList()
				properties.addAll(parentProperties)
			}
			// Add properties herited from grand parents
			for (parent in parentRolesSorted) {
				val parentProperties = buildProperties(parent, false, true, false, ignoredPropertyTypes).toList()
				properties.addAll(parentProperties)
			}
		}
		if (includeDefault) { // Get the default role and add property if not overridden in child role
			roleDAO.getRoleByName(Roles.DEFAULT_ROLE_NAME)?.let {
				for (defaultProp in it.properties) {
					if (!ignoredPropertyTypes.contains(defaultProp.type)) {
						properties.add(defaultProp)
					}
				}
			}
		}
		emitAll(properties.asFlow())
	}

	override fun getAscendantRoles(principalId: String): Flow<Role> = flow<Role> {
		getPrincipal(principalId)?.let { buildAscendantRoles(it, mutableSetOf()).collect { emit(it) } }
	}

	protected fun buildAscendantRoles(principal: Principal?, ignoredRoles: MutableSet<Role>): Flow<Role> = flow {
		val roles = mutableSetOf<Role>()
		principal?.let {
			if (principal is Role) {
				ignoredRoles.add(principal)
			}
			principal.getParents().let {
				getParents(principal).onEach {
					it.let {
						if (!ignoredRoles.contains(it)) {
							roles.add(it)
							roles.addAll(buildAscendantRoles(it, ignoredRoles).toList())
						}
					}
				}.collect()
			}
		}
		emitAll(roles.asFlow())
	}

	companion object {
		// Bit indexes
		protected var DEPENDENCY_INCLUDE_DIRECT = 1 shl 1
		protected var DEPENDENCY_INCLUDE_HERITED = 1 shl 2
	}
}
